google-workloads-to-champsim/README.md

```diff
@@ -0,0 +1,95 @@
+# Google Workload Traces to Champsim Wrapper Script
+
+## Compile
+Use the following command to compile: 
+
+`g++ google-workload-to-champsim.cpp -lz` 
+
+## Run
+Use the following command to run the executable:
+
+`./a.out <trace>`
+
+`<trace>` should be a google workload trace file in compressed form with `.gz` extension. Google workload traces can be obtained [here](https://dynamorio.org/google_workload_traces.html).
+
+A new trace file compatible with champsim will be created. "champsim-" will be prefixed to the original trace file's name. This trace file will also be in compressed form with `.gz` extension.
+
+## About Wrapper Script
+
+### Google Workload Trace Format
+Google Workload traces are stored as records, each record being of type `trace_entry_t` and size 12 bytes. Following is the structure of `trace_entry_t`:
```

Comment:
For the conversion, we would highly recommend using our libraries which read the disk format into a record format suitable for analysis and simulation (the memref_t type).  The disk format is subject to change without considering that a compatibility change, as it is not meant to be a public relied-upon interface.  Furthermore, there are numerous complexities built into the disk format, such as including instruction encodings or physical address values only the first time observed, which are abstracted away by the libraries.  Trying to directly convert the disk format would require duplicating the handling of those complexities.  Best to work at the recommended and documented higher level for simplicity and to avoid breaking changes.

Thus, we would request to not take this approach and to instead create an analysis tool in our framework and convert from the memref_t level.  There are numerous changes at the disk level that have been made in the last year which break this converter as-is: it won't work on today's traces.

Reply:
Going further, for our user-mode software-thread-oriented traces, we would recommend building a dynamic converter from memref_t, rather than an offline static converter, for use with our dynamic scheduler which re-schedules our traces onto the simulated virtual cores.  This solves several issues: it allows for mixing multiple workloads to model multi-tenant datacenters yet per-workload traces; it deflates the context switch inflation added due to tracing overhead; it (will) provide some functionality typically missing from trace-based simulation such as speculative path support by pulling instructions from prior points in the trace or using other schemes.

Reply:
What do you suggest we do instead? Would this fall under the category of analysis tools, as described [here](https://dynamorio.org/sec_drcachesim_newtool.html)? Or is there a different way to get to a `memref_t`?

Reply:
Yes, an analysis tool.  If ChampSim files are per-software-thread, a parallel tool with software threads as the parallel shards would work.  You could start with an offline conversion to ChampSim files in the analysis tool.  The same analysis tool code could be run with the shards as virtual cores (possibly with a dynamic reschedule of the software threads) to produce per-hardware-thread trace files.